### PR TITLE
Added char buffer

### PR DESCRIPTION
**Pull request by niloo9876**
- Had to change the MCU buffer to 30MB for the program to fit into MRAM
- Added a char cache as well as char MCU output buffer (The naming is not the best, but I can change that later)
- The code might not work with the CPU (All decoding on the host), I'll work on changing the CPU code so we can do comparison testing
- Code tested, but I have not done any performance measures.

### CODE REVIEW

**Review by mkdashti — CHANGES_REQUESTED** [review]
Looks good, except you have to make one additional change in jpeg-host.c since with the current changes, you have to copy "MCU_out_buffer" from DPU not "MCU_buffer".
[jpeg-host.c line 96](https://github.com/UBC-ECE-Sasha/PIM-JPEG/blob/c3335b46068f85d949992cfc57aa344f1e72f95e/src/jpeg-host.c#L96)

So, the change would be something like this:
`DPU_ASSERT(dpu_push_xfer(dpus, DPU_XFER_FROM_DPU, "MCU_out_buffer", 0, sizeof(char) * largest_pixel_count * 3,
                           DPU_XFER_ASYNC));`

Other than this, I think this looks great and should reduce copying time from DPU by half!

File: src/bmp.c [hunk 1]
```diff
@@ -103,11 +103,12 @@ static int write_bmp(const char *filename, uint32_t image_width, uint32_t image_
 
 int write_bmp_cpu(const char *filename, uint32_t image_width, uint32_t image_height, uint32_t image_padding,
                   uint32_t mcu_width, short *MCU_buffer) {
-  return write_bmp(filename, image_width, image_height, image_padding, mcu_width, MCU_buffer, 0);
+  // TODO FIX the code for CPU
```
Comment (jnider) [inline src/bmp.c:106]: just add an issue for this; it's easier to find & remember later

File: src/dpu/dpu-jpeg-decode.c [hunk 2]
```diff
@@ -4,10 +4,12 @@
 
 #include "dpu-jpeg.h"
 
-__mram_noinit short MCU_buffer[NR_TASKLETS][16776960 / NR_TASKLETS];
+__mram_noinit short MCU_buffer[NR_TASKLETS][MEGABYTE(15) / NR_TASKLETS]; // 32MB
+__mram_noinit char MCU_out_buffer[MEGABYTE(15)]; // 16MB
 
 #define PREWRITE_SIZE 768
 __dma_aligned short MCU_buffer_cache[NR_TASKLETS][PREWRITE_SIZE];
+__dma_aligned char MCU_out_cache[NR_TASKLETS][PREWRITE_SIZE];
```
Comment (jnider) [inline src/dpu/dpu-jpeg-decode.c:12]: This isn't really a cache, but not important for now

File: src/dpu/dpu-jpeg-decode.c [hunk 3]
```diff
@@ -363,7 +365,8 @@ void inverse_dct_convert(JpegDecompressor *d) {
 
           ycbcr_to_rgb_pixel(d, cache_index, y, x);
 
-          mram_write(&MCU_buffer_cache[d->tasklet_id][cache_index], &MCU_buffer[0][mcu_index], MCU_READ_WRITE_SIZE1);
+          // After this all the values are between 0 and 255
+          mram_write(&MCU_out_cache[d->tasklet_id][cache_index], &MCU_out_buffer[mcu_index], MCU_READ_WRITE_SIZE1 >> 1);
```
Comment (jnider) [inline src/dpu/dpu-jpeg-decode.c:369]: I never liked those MCU_READ_WRITE_SIZE macros - it should be defined as: MCU_size * 3
(also not important for now)

File: src/dpu/dpu-jpeg-decode.c [hunk 4]
```diff
@@ -521,7 +524,8 @@ static void ycbcr_to_rgb_pixel(JpegDecompressor *d, int cache_index, int v, int
   int max_v = jpegInfo.max_v_samp_factor;
   int max_h = jpegInfo.max_h_samp_factor;
 
-  // Iterating from bottom right to top left because otherwise the pixel data will get overwritten
+  // Iterating from bottom right to top left because otherwise the pixel data will get overwritten 
+  // due to the horizantal and vertical sampling factors
```
Comment (jnider) [inline src/dpu/dpu-jpeg-decode.c:528]: horizantal -> horizontal

File: src/dpu/dpu-jpeg-decode.c [hunk 5]
```diff
@@ -4,8 +4,8 @@
 
 #include "dpu-jpeg.h"
 
-__mram_noinit short MCU_buffer[NR_TASKLETS][MEGABYTE(15) / NR_TASKLETS]; // 32MB
-__mram_noinit char MCU_out_buffer[MEGABYTE(15)]; // 16MB
+__mram_noinit short MCU_buffer[NR_TASKLETS][MEGABYTE(15) / NR_TASKLETS]; // 30MB
+__mram_noinit char MCU_out_buffer[MEGABYTE(15)]; // 15MB
```
Comment (jnider) [inline src/dpu/dpu-jpeg-decode.c:8]: In a perfect world, you would just 'git squash' these commits together

File: src/jpeg-host.c [hunk 6]
```diff
@@ -183,9 +183,9 @@ static int dpu_main(struct jpeg_options *opts, host_results *results) {
   }
 
   dpu_output_t *dpu_outputs = calloc(dpu_count, sizeof(dpu_output_t));
-  short **MCU_buffer = malloc(sizeof(short *) * dpu_count);
+  char **MCU_buffer = malloc(sizeof(char *) * dpu_count);
   for (dpu_id = 0; dpu_id < dpu_count; dpu_id++) {
-    MCU_buffer[dpu_id] = malloc(sizeof(short) * 87380 * 3 * 64);
+    MCU_buffer[dpu_id] = malloc(sizeof(char) * 87380 * 3 * 64);
```
Comment (jnider) [inline src/jpeg-host.c:188]: this still has magic numbers - we'll have to get rid of that eventually

**Review by jnider — APPROVED** [review]

**Review by mkdashti — APPROVED** [review]
Looks good.

**Pull request merged**